common/message_algos/src/grouping/message_group.rs

```diff
@@ -0,0 +1,36 @@
+use crate::{Envelope, Message};
+
+/// A group of messages (or message envelopes). Guaranteed to contain at least one message.
+#[derive(Debug)]
+pub struct MessageGroup<T: Message> {
+    messages: Vec<Envelope<T>>,
```

Comment:
Storing the message envelopes in a `HashSet` or a `HashMap` instead of a `Vec` would have made the duplication handling less costly, avoiding the need to iterate over all the messages.